src/Microsoft.ApplicationInsights.AspNetCore/TelemetryInitializers/AzureAppServiceRoleNameFromHostNameHeaderInitializer.cs

```diff
@@ -0,0 +1,149 @@
+        {
+            this.httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
+
+            try
+            {
+                var result = Environment.GetEnvironmentVariable(WebAppHostNameEnvironmentVariable);
+
+                if (!string.IsNullOrEmpty(result) && result.EndsWith(this.WebAppSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    result = result.Substring(0, result.Length - this.WebAppSuffix.Length);
+                }
+
+                this.roleName = result;
+            }
+            catch (Exception ex)
+            {
+                AspNetCoreEventSource.Instance.LogAzureAppServiceRoleNameFromHostNameHeaderInitializerWarning(ex.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets suffix of website name. This must be changed when running in non public Azure region.
+        /// Default value (Public Cloud):  ".azurewebsites.net"
+        /// For US Gov Cloud: ".azurewebsites.us"
+        /// For Azure Germany: ".azurewebsites.de".
+        /// </summary>
+        public string WebAppSuffix { get; set; } = ".azurewebsites.net";
+
+        /// <summary>
+        /// Populates RoleName from the request telemetry associated with the http context.
+        /// If RoleName is empty on the request telemetry, it'll be updated as well so that other telemetry
+        /// belonging to the same requests gets it from request telemetry, without having to parse headers again.
+        /// </summary>
+        /// <remarks>
+        /// RoleName is attempted from every incoming request as opposed to doing this periodically. This is
+        /// done to ensure every request (and associated telemetry) gets the correct RoleName during slot swap.
+        /// </remarks>
+        /// <param name="telemetry">The telemetry item for which RoleName is to be set.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+                {
+                    // RoleName is already populated.
+                    return;
+                }
+
+                string roleName = string.Empty;
+                var context = this.httpContextAccessor.HttpContext;
+
+                if (context != null)
+                {
+                    lock (context)
+                    {
+                        var request = context.Features.Get<RequestTelemetry>();
+
+                        if (request != null)
+                        {
+                            if (string.IsNullOrEmpty(request.Context.Cloud.RoleName))
+                            {
+                                roleName = this.GetRoleNameFromHeader(context);
+                                if (!string.IsNullOrEmpty(roleName))
```

Comment:
I think this could be simplified by changing it to a `if (this.TryGetRoleNameFromHeader(context, out roleName)` so you won't need to repeat the IsNullOrEmpty check.